File: day5/python/solution.py
import numpy as np

def get_pairs(start, end, diag):
    x0 = start[0]
    x1 = end[0]
    y0 = start[1]
    y1 = end[1]
    if (x0 < x1):
        xs = [x for x in range(x0, x1 + 1, 1)]
    else:
        xs = [x for x in range(x0, x1 - 1, -1)]
    if (y0 < y1):
        ys = [y for y in range(y0, y1 + 1, 1)]
    else:
        ys = [y for y in range(y0, y1 - 1, -1)]
    if (x0 == x1 or y0 == y1):
        return [(x,y) for x in xs for y in ys]
    else:
        if not diag:
            return []
        result = []
        for i in range(len(xs)):
            result.append((xs[i], ys[i]))
        return result

def print_example(dict):
    matrix = np.zeros((10, 10))
    for pair in dict:
        matrix[pair[1]][pair[0]] = dict[pair]
    print(matrix)

with open("../input.txt") as file:
    input = file.read().splitlines()
    lines = map(lambda range: range.split(' -> '), input)


def main():
    board = {}
    for line in lines:
        start = tuple(list(map(int, line[0].split(','))))
        end = tuple(list(map(int, (line[1].split(',')))))
        pairs = get_pairs(start, end, True)
        for pair in pairs:
            if not pair in board:
                board[pair] = 0
            board[pair] += 1

    count = 0
    for points in board.values():
        if points >= 2:
            count += 1
    print(count)
    
main()

File: day5/haskell/solution.hs
module Main where
import Data.List.Split
import qualified Data.HashMap as HM

type Point = (Int, Int)
data Line = Line Point Point
    deriving (Show, Eq)

-- | Gives all poitns on a line, 
--   if second argument is false it ignores diagonals
getPoints :: Line -> Bool -> [Point]
getPoints (Line (x0, y0) (x1, y1)) includeDiags
    | x0 == x1 || y0 == y1 = [(x, y) | x <- xs, y <- ys]
    | includeDiags = [(xs !! i, ys !! i) | i <- [0..(length xs - 1)]]
    | otherwise = []
        where
            xs = if x0 < x1 then [x0..x1] else [x0,x0-1..x1]
            ys = if y0 < y1 then [y0..y1] else [y0,y0-1..y1]


nIntersect :: [Point] -> HM.Map Point Int
nIntersect ps = calc ps HM.empty
    where
        calc [] map = map
        calc (p:ps) map = calc ps (HM.insertWith (\n o -> o+1) p 1 map)

main :: IO ()
main = do
    input <- readFile "../input.txt"
    -- Unreadable but fun with oneliner, can be split up
    let points = concatMap ((`getPoints` True) . (\ps -> Line (head ps) (last ps)) 
            . map ((\s -> (read $ head s, read $ last s)) . splitOn ",")
                . filter (/= "->") . words) (lines input)
    print $ length $ HM.filter (>= 2) $ nIntersect points